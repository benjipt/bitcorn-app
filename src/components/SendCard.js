import React, { useState } from 'react'

const baseURL = 'https://jobcoin.gemini.com/greyhound-abruptly/api/'

export default function SendCard({ address, getData }) {

    // STATE HOOKS
    const [ nsfError, setNsfError ] = useState(false)
    const [ inputValue, setInputValue ] = useState({
        toAddress: '',
        amount: ''
    })

    // Destructured properties of inputValue state hook
    const { toAddress, amount } = inputValue

    const handleChange = e => {
        const { id, value } = e.currentTarget
        setInputValue({ ...inputValue, [id]: value })
    }

    const handleSubmit = e => {
        e.preventDefault()
        fetch(baseURL + 'transactions', {
            method: 'POST',
            body: JSON.stringify({
                fromAddress: address,
                toAddress: toAddress,
                amount: amount
            }),
            headers: {
                'Content-Type': 'application/json'
            }
        }).then(res => {
            if (res.status === 422) {
                setNsfError(true)
            } else {
                setInputValue({
                    toAddress: '',
                    amount: ''
                })
                getData(address)
            }
        })
    }

    return (
        <div className="container mt-5 p-4 border-black rounded card-custom" data-testid="SendCard-1">
            <div className="text-center">
                <h5>Send JobCoin</h5>
            </div>
            <hr></hr>
            <form onSubmit={ handleSubmit } data-testid="SendCard-2">
                <div className="mb-3">
                    <label htmlFor="toAddress" className="form-label">Destination Address</label>
                    <input onChange={ handleChange } type="text" className="form-control" id="toAddress" data-testid="SendCard-input-1" value={ toAddress } />
                </div>
                <div className="mb-3">
                    <label htmlFor="amount" className="form-label">Amount to Send</label>
                    <input onChange={ handleChange } type="text" className="form-control" id="amount" data-testid="SendCard-input-2" value={ amount } />
                    { nsfError &&
                        <div id="nsfError" className="form-text">You do not have enough funds to send this amount.</div> }
                </div>
                <button type="submit" className="btn btn-primary">Send</button>
            </form>
        </div>
    )
}
