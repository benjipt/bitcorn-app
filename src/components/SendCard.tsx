import { useState, ChangeEvent, SyntheticEvent } from 'react';
import { postTransaction } from '../utils/api';

interface SendCardProps {
  address: string;
  getData: (address: string) => void;
}

export default function SendCard({ address, getData }: SendCardProps) {
  interface InputValue {
    toAddress: string;
    amount: string;
  }
  // STATE HOOKS
  const [nsfError, setNsfError] = useState(false);
  const [inputValue, setInputValue] = useState<InputValue>({
    toAddress: '',
    amount: '',
  });

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const { id, value } = e.currentTarget;
    setInputValue({ ...inputValue, [id]: value });
  };

  const { toAddress, amount } = inputValue;
  const handleSubmit = async (e: SyntheticEvent<HTMLFormElement>) => {
    e.preventDefault();
    try {
      const res = await postTransaction(address, toAddress, amount);
      if (res.status === 422) {
        setNsfError(true);
      } else {
        setInputValue({
          toAddress: '',
          amount: '',
        });
        getData(address);
      }
    } catch (error) {
      // Handle errors, such as network issues or API errors
      console.error('Error submitting transaction:', error);
    }
  };

  return (
    <div
      className='container mt-5 p-4 border-black rounded card-custom'
      data-testid='SendCard-1'>
      <div className='text-center'>
        <h5>Send JobCoin</h5>
      </div>
      <hr></hr>
      <form onSubmit={handleSubmit}>
        <div className='mb-3'>
          <label htmlFor='toAddress' className='form-label'>
            Destination Address
          </label>
          <input
            onChange={handleChange}
            type='text'
            className='form-control'
            id='toAddress'
            data-testid='input-1'
            value={toAddress}
          />
        </div>
        <div className='mb-3'>
          <label htmlFor='amount' className='form-label'>
            Amount to Send
          </label>
          <input
            onChange={handleChange}
            type='text'
            className='form-control'
            id='amount'
            data-testid='input-2'
            value={amount}
          />
          {nsfError && (
            <div id='nsfError' className='form-text' data-testid='error-1'>
              You do not have enough funds to send this amount.
            </div>
          )}
        </div>
        <button
          type='submit'
          className='btn btn-primary'
          data-testid='submit-1'>
          Send
        </button>
      </form>
    </div>
  );
}
