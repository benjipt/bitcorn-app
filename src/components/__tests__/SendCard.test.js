import { render, cleanup, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer'
import SendCard from '../SendCard'


describe('SendCard', () => {

    beforeEach(() => {
        render(<SendCard/>)
    })
    
    afterEach(() => {
        cleanup()
    })

    test('should render SendCard component', () => {
        screen.getByTestId('SendCard-1')
    })

    describe('Form inputs', () => {
        
        test('should render input fields', () => {
            screen.getByTestId('SendCard-input-1')
            screen.getByTestId('SendCard-input-2')
        })
    
        test('handleChange updates controlled input values', () => {
            // Accessing Input fields
            const toAddressInput = screen.getByTestId('SendCard-input-1')
            const amountInput = screen.getByTestId('SendCard-input-2')
            // Typing input values
            userEvent.type(toAddressInput, 'Bob')
            userEvent.type(amountInput, '24.5')
            // Tests for expected input values
            expect(toAddressInput).toHaveValue('Bob')
            expect(amountInput).toHaveValue('24.5')
        })
    })

    describe('Form submit', () => {

        test('should render submit button', () => {
            screen.getByText('Send')
        })

    })

    test('matches snapshot', () => {
        const tree = renderer.create(<SendCard/>).toJSON()
        expect(tree).toMatchSnapshot()
    })

})