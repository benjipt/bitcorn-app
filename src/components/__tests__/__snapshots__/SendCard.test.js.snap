// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendCard matches snapshot 1`] = `
<div
  className="container mt-5 p-4 border-black rounded card-custom"
  data-testid="SendCard-1"
>
  <div
    className="text-center"
  >
    <h5>
      Send JobCoin
    </h5>
  </div>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="toAddress"
      >
        Destination Address
      </label>
      <input
        className="form-control"
        data-testid="SendCard-input-1"
        id="toAddress"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="amount"
      >
        Amount to Send
      </label>
      <input
        className="form-control"
        data-testid="SendCard-input-2"
        id="amount"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <button
      className="btn btn-primary"
      data-testid="SendCard-submit-1"
      type="submit"
    >
      Send
    </button>
  </form>
</div>
`;
